actor/builtin/miner/sector_set.go

```diff
@@ -0,0 +1,74 @@
+package miner
+
+import (
+	"strconv"
+
+	"github.com/filecoin-project/go-filecoin/types"
+	"github.com/filecoin-project/go-filecoin/vm/errors"
+)
+
+// Note that this type's interface is a bit of a placeholder.  We expect to
+// move to something like actor.Lookup eventually which will manage its own
+// hash linked storage.  See #2887.  This will require some interface changes.
+
+// SectorSet is a collection of sector commitments indexed by their integer
+// sectorID.  Due to a bug in refmt, the sector id-keys need to be stringified.
+// See also: https://github.com/polydawn/refmt/issues/35.
+type SectorSet map[string]types.Commitments
+
+// NewSectorSet initializes a SectorSet with no entries.
+func NewSectorSet() SectorSet {
+	return make(map[string]types.Commitments)
+}
+
+// Has returns true if the SectorSet is already tracking id, else false.
+func (ss SectorSet) Has(id uint64) bool {
+	_, ok := ss[idStr(id)]
+	return ok
+}
+
+// Add updates the SectorSet to include the new commitment at the given id.
+func (ss SectorSet) Add(id uint64, comms types.Commitments) {
+	ss[idStr(id)] = comms
+}
+
+// Get returns the commitment at the given id and a bool indicating success.
+func (ss SectorSet) Get(id uint64) (types.Commitments, bool) {
+	comms, ok := ss[idStr(id)]
+	return comms, ok
+}
+
+// Drop removes all the provided sectorIDs from the collection, erroring if any
+// do not exist in the SectorSet.
+func (ss SectorSet) Drop(ids []uint64) error {
+	for _, id := range ids {
+		if _, ok := ss[idStr(id)]; !ok {
+			return Errors[ErrInvalidSector]
+		}
+		delete(ss, idStr(id))
+	}
+	return nil
+}
+
+// IDs returns all of the sectorIDs in the sectorset.  They are not sorted.
+func (ss SectorSet) IDs() ([]uint64, error) {
+	var ids []uint64
+	for idStr := range ss {
+		id, err := str2ID(idStr)
+		if err != nil {
+			return nil, errors.RevertErrorWrap(err, "corrupt sectorset id")
```

Comment:
I would expect this wrapping to be done at the VM boundary: code inside the actor can either succeed or fail, and any failure will revert the transaction but not be a "fault" error. The actor code in general should not depend on these error abstractions.

Reply:
Am I correctly reading this as a suggestion that we not call `errors.NewRevertError` or `errors.NewFaultError` from within actor code?  If so this is a big technical debt concern because we do this very commonly in actor code.

Am I misreading this?  If so could you please clarify the concrete change you'd like to see?

Reply:
The way the VM is set up, it requires actors to declare errors to be reverts or faults. This makes some sense because this determination can't be made externally, but it would be nice if faults had to be wrapped and any non-fault error was treated as a revert. I think this system was set up for safety (forcing the developer to choose).

Reply:
!! @ZenGround0 you interpreted correctly but I misunderstood the current convention and am now very alarmed by it.

> set up for safety

But it's not! Nothing forces the developer to choose because the methods are declared to return simply `error`. A quick grep shows at least five calls to methods like `errors.{New|Wrap|Errorf}` in actor code. I would bet large amounts of money that there are errors returned from code called by actors that are not wrapped. 

Also, this is supposed to be a VM, how can anything from inside here be intentionally causing a node to shutdown? If such a case is part of message execution, a node could only crash-loop until the message expired from all pools. That's a horrible DOS vector.

Reply:
It looks like "faults" are not handled as harshly as they're documented to be, so a node won't shutdown. I've filed #3009 to clear this all up (later).